File: service/src/test/java/no/nav/melosys/service/dokument/brev/mapper/AvslagYrkesaktivMapperTest.java
package no.nav.melosys.service.dokument.brev.mapper;

import java.time.LocalDate;
import java.util.*;
import jakarta.xml.bind.JAXBException;

import no.nav.dok.brevdata.felles.v1.navfelles.Kontaktinformasjon;
import no.nav.dok.melosysbrev._000081.Fag;
import no.nav.dok.melosysbrev.felles.melosys_felles.FellesType;
import no.nav.dok.melosysbrev.felles.melosys_felles.MelosysNAVFelles;
import no.nav.melosys.domain.*;
import no.nav.melosys.domain.adresse.StrukturertAdresse;
import no.nav.melosys.domain.avklartefakta.AvklartVirksomhet;
import no.nav.melosys.domain.kodeverk.*;
import no.nav.melosys.domain.kodeverk.begrunnelser.Utsendt_arbeidstaker_begrunnelser;
import no.nav.melosys.domain.kodeverk.begrunnelser.Anmodning_begrunnelser;
import no.nav.melosys.domain.kodeverk.begrunnelser.Avslag_anmodning_begrunnelser;
import no.nav.melosys.domain.kodeverk.yrker.Yrkesaktivitetstyper;
import no.nav.melosys.domain.mottatteopplysninger.MottatteOpplysninger;
import no.nav.melosys.domain.mottatteopplysninger.Soeknad;
import no.nav.melosys.domain.mottatteopplysninger.data.arbeidssteder.FysiskArbeidssted;
import no.nav.melosys.service.dokument.brev.BrevDataAvslagYrkesaktiv;
import no.nav.melosys.service.dokument.brev.BrevbestillingDto;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.xml.sax.SAXException;

import static no.nav.melosys.service.dokument.brev.BrevDataTestUtils.lagAnmodningsperiodeSvarAvslag;
import static no.nav.melosys.service.dokument.brev.BrevDataUtils.lagKontaktInformasjon;
import static no.nav.melosys.service.dokument.brev.BrevDataUtils.lagNorskPostadresse;
import static no.nav.melosys.service.dokument.brev.mapper.BrevMappingTestUtils.lagNAVFelles;
import static no.nav.melosys.service.dokument.brev.mapper.felles.VilkaarbegrunnelseFactoryTest.lagAlleVilkaarBegrunnelser;
import static org.assertj.core.api.Assertions.assertThat;

class AvslagYrkesaktivMapperTest {

    private FellesType fellesType;
    private Behandling behandling;
    private MelosysNAVFelles navFelles;

    @BeforeEach
    public void setUp() {
        fellesType = new FellesType();
        fellesType.setFagsaksnummer("MELTEST-1");

        navFelles = lagNAVFelles();
        navFelles.getMottaker().setMottakeradresse(lagNorskPostadresse());
        Kontaktinformasjon kontaktinformasjon = lagKontaktInformasjon();
        navFelles.setKontaktinformasjon(kontaktinformasjon);

        behandling = BehandlingTestFactory.builderWithDefaults().build();
        Fagsak fagsak = FagsakTestFactory.lagFagsak();
        behandling.setFagsak(fagsak);
    }

    @Test
    void mapTilBrevXML() throws JAXBException, SAXException {

        FysiskArbeidssted fysiskArbeidssted = new FysiskArbeidssted("NO", new StrukturertAdresse());

        Soeknad soeknad = new Soeknad();
        soeknad.arbeidPaaLand.setFysiskeArbeidssteder(new ArrayList<>());
        soeknad.arbeidPaaLand.getFysiskeArbeidssteder().add(fysiskArbeidssted);

        MottatteOpplysninger mottatteOpplysninger = new MottatteOpplysninger();
        mottatteOpplysninger.setMottatteOpplysningerData(soeknad);
        behandling.setMottatteOpplysninger(mottatteOpplysninger);

        Behandlingsresultat resultat = new Behandlingsresultat();

        Lovvalgsperiode lovvalgsperiode = new Lovvalgsperiode();
        lovvalgsperiode.setLovvalgsland(Land_iso2.NO);
        lovvalgsperiode.setFom(LocalDate.now());
        lovvalgsperiode.setTom(LocalDate.now());
        resultat.setLovvalgsperioder(Collections.singleton(lovvalgsperiode));

        resultat.setVilkaarsresultater(new HashSet<>());

        Vilkaarsresultat vilkaarsresultat12_1 = lagVilkaarsresultat(Vilkaar.FO_883_2004_ART12_1, false, Utsendt_arbeidstaker_begrunnelser.IKKE_VESENTLIG_VIRKSOMHET);
        resultat.getVilkaarsresultater().add(vilkaarsresultat12_1);

        Vilkaarsresultat vilkaarsresultat12_2 = lagVilkaarsresultat(Vilkaar.FO_883_2004_ART12_2, false);
        resultat.getVilkaarsresultater().add(vilkaarsresultat12_2);

        Vilkaarsresultat vilkaarsresultat16_1 = lagVilkaarsresultat(Vilkaar.FO_883_2004_ART16_1, false,
            Avslag_anmodning_begrunnelser.OVER_5_AAR,
            Avslag_anmodning_begrunnelser.SOEKT_FOR_SENT,
            Avslag_anmodning_begrunnelser.SAERLIG_AVSLAGSGRUNN);
        vilkaarsresultat16_1.setBegrunnelseFritekst("Fritekst");

        BrevDataAvslagYrkesaktiv brevData = new BrevDataAvslagYrkesaktiv(new BrevbestillingDto(), "Z999999");
        brevData.setArbeidsland(Landkoder.AT.getBeskrivelse());
        brevData.setHovedvirksomhet(new AvklartVirksomhet("Test AS", null, null, Yrkesaktivitetstyper.LOENNET_ARBEID));
        brevData.setAnmodningsperiodeSvar(new AnmodningsperiodeSvar());
        brevData.setYrkesaktivitet(Yrkesaktivitetstyper.LOENNET_ARBEID);
        brevData.setArt16Vilkaar(vilkaarsresultat16_1);

        AvslagYrkesaktivMapper spy = Mockito.spy(new AvslagYrkesaktivMapper());
        String xml = spy.mapTilBrevXML(fellesType, navFelles, behandling, resultat, brevData);

        assertThat(xml).matches("(?s)\\<\\?xml version=\"\\d\\.\\d+\" .*>\n.*");
    }

    @Test
    void mapTilBrevXML_medOppfyltArt16OgAnmodningsperiode_brukerAnmodningsperiode() {
        AvslagYrkesaktivMapper spy = Mockito.spy(new AvslagYrkesaktivMapper());

        BrevDataAvslagYrkesaktiv brevData = new BrevDataAvslagYrkesaktiv(new BrevbestillingDto(), "Z999999");
        brevData.setArbeidsland(Landkoder.ES.getBeskrivelse());
        brevData.setHovedvirksomhet(new AvklartVirksomhet("Test AS", null, null, Yrkesaktivitetstyper.LOENNET_ARBEID));
        brevData.setAnmodningsperiodeSvar(lagAnmodningsperiodeSvarAvslag());
        brevData.setYrkesaktivitet(Yrkesaktivitetstyper.LOENNET_ARBEID);

        Vilkaarsresultat vilkår16_1_oppfylt = lagVilkaarsresultat(Vilkaar.FO_883_2004_ART16_1, true, Anmodning_begrunnelser.ERSTATTER_EN_ANNEN_UNDER_5_AAR);
        brevData.setArt16Vilkaar(vilkår16_1_oppfylt);

        Behandlingsresultat resultat = lagBehandlingsresultat();
        Vilkaarsresultat vilkår12_1_avslått = lagVilkaarsresultat(Vilkaar.FO_883_2004_ART12_1, false, Utsendt_arbeidstaker_begrunnelser.IKKE_VESENTLIG_VIRKSOMHET);
        resultat.getVilkaarsresultater().add(vilkår12_1_avslått);

        String xml = spy.mapTilBrevXML(fellesType, navFelles, behandling, resultat, brevData);
        assertThat(xml).matches("(?s)\\<\\?xml version=\"\\d\\.\\d+\" .*>\n.*");
    }

    @Test
    void mapTilBrevXml_kanMappeAlleKodeverksverdierForArt16_1_avslag() throws Exception {
        AvslagYrkesaktivMapper spy = Mockito.spy(new AvslagYrkesaktivMapper());
        BrevDataAvslagYrkesaktiv brevdata = new BrevDataAvslagYrkesaktiv(new BrevbestillingDto(), "");
        Set<VilkaarBegrunnelse> begrunnelser = lagAlleVilkaarBegrunnelser(Avslag_anmodning_begrunnelser.class);
        for (VilkaarBegrunnelse begrunnelse : begrunnelser) {
            Vilkaarsresultat vilkaarsresultat = new Vilkaarsresultat();
            vilkaarsresultat.setBegrunnelser(Collections.singleton(begrunnelse));
            vilkaarsresultat.setBegrunnelseFritekst("Fritekst");
            brevdata.setArt16Vilkaar(vilkaarsresultat);
            spy.mapArt161Avslag(new Fag(), brevdata);
        }
    }

    private Vilkaarsresultat lagVilkaarsresultat(Vilkaar vilkaar, boolean oppfylt, Kodeverk... vilkårbegrunnelser) {
        Vilkaarsresultat vilkaarsresultat = new Vilkaarsresultat();
        vilkaarsresultat.setOppfylt(oppfylt);
        vilkaarsresultat.setVilkaar(vilkaar);
        vilkaarsresultat.setBegrunnelser(new HashSet<>());
        for (Kodeverk begrunnelseKode : vilkårbegrunnelser) {
            VilkaarBegrunnelse begrunnelse = new VilkaarBegrunnelse();
            begrunnelse.setKode(begrunnelseKode.getKode());
            vilkaarsresultat.getBegrunnelser().add(begrunnelse);
        }
        return vilkaarsresultat;
    }

    private Behandlingsresultat lagBehandlingsresultat() {
        Behandlingsresultat resultat = new Behandlingsresultat();

        Lovvalgsperiode lovvalgsperiode = new Lovvalgsperiode();
        lovvalgsperiode.setLovvalgsland(Land_iso2.NO);
        lovvalgsperiode.setFom(LocalDate.now());
        lovvalgsperiode.setTom(LocalDate.now());
        resultat.setLovvalgsperioder(Collections.singleton(lovvalgsperiode));
        resultat.setVilkaarsresultater(new HashSet<>());
        return resultat;
    }

}

File: service/src/test/kotlin/no/nav/melosys/service/dokument/brev/mapper/AvslagYrkesaktivMapperTest.kt
package no.nav.melosys.service.dokument.brev.mapper

import io.kotest.matchers.string.shouldMatch
import io.mockk.spyk
import no.nav.dok.melosysbrev._000081.Fag
import no.nav.dok.melosysbrev.felles.melosys_felles.FellesType
import no.nav.dok.melosysbrev.felles.melosys_felles.MelosysNAVFelles
import no.nav.melosys.domain.AnmodningsperiodeSvar
import no.nav.melosys.domain.Behandling
import no.nav.melosys.domain.Behandlingsresultat
import no.nav.melosys.domain.avklartefakta.AvklartVirksomhet
import no.nav.melosys.domain.begrunnelse
import no.nav.melosys.domain.forTest
import no.nav.melosys.domain.kodeverk.Land_iso2
import no.nav.melosys.domain.kodeverk.Landkoder
import no.nav.melosys.domain.kodeverk.Vilkaar
import no.nav.melosys.domain.kodeverk.begrunnelser.Anmodning_begrunnelser
import no.nav.melosys.domain.kodeverk.begrunnelser.Avslag_anmodning_begrunnelser
import no.nav.melosys.domain.kodeverk.begrunnelser.Utsendt_arbeidstaker_begrunnelser
import no.nav.melosys.domain.kodeverk.yrker.Yrkesaktivitetstyper
import no.nav.melosys.domain.lovvalgsperiode
import no.nav.melosys.domain.mottatteOpplysninger
import no.nav.melosys.domain.mottatteopplysninger.soeknad
import no.nav.melosys.domain.vilkaarsresultat
import no.nav.melosys.domain.vilkaarsresultatForTest
import no.nav.melosys.service.dokument.brev.BrevDataAvslagYrkesaktiv
import no.nav.melosys.service.dokument.brev.BrevDataTestUtils.lagAnmodningsperiodeSvarAvslag
import no.nav.melosys.service.dokument.brev.BrevDataUtils.lagKontaktInformasjon
import no.nav.melosys.service.dokument.brev.BrevDataUtils.lagNorskPostadresse
import no.nav.melosys.service.dokument.brev.BrevbestillingDto
import no.nav.melosys.service.dokument.brev.mapper.BrevMappingTestUtils.lagNAVFelles
import no.nav.melosys.service.dokument.brev.mapper.felles.VilkaarbegrunnelseFactoryTest
import org.junit.jupiter.api.Test
import java.time.LocalDate

class AvslagYrkesaktivMapperTest {

    // Brev-types use .apply since they don't have forTest DSL
    private val fellesType = FellesType().apply {
        fagsaksnummer = "MELTEST-1"
    }

    private val navFelles = lagNAVFelles().apply {
        mottaker.mottakeradresse = lagNorskPostadresse()
        kontaktinformasjon = lagKontaktInformasjon()
    }

    @Test
    fun `mapTilBrevXML genererer gyldig XML`() {
        val behandling = Behandling.forTest {
            mottatteOpplysninger {
                soeknad {
                    fysiskeArbeidssted {
                        landkode = "NO"
                    }
                }
            }
        }

        val vilkaarsresultat16_1 = vilkaarsresultatForTest {
            vilkaar = Vilkaar.FO_883_2004_ART16_1
            isOppfylt = false
            begrunnelse(Avslag_anmodning_begrunnelser.OVER_5_AAR.kode)
            begrunnelse(Avslag_anmodning_begrunnelser.SOEKT_FOR_SENT.kode)
            begrunnelse(Avslag_anmodning_begrunnelser.SAERLIG_AVSLAGSGRUNN.kode)
            begrunnelseFritekst = "Fritekst"
        }

        val resultat = Behandlingsresultat.forTest {
            lovvalgsperiode {
                lovvalgsland = Land_iso2.NO
                fom = LocalDate.now()
                tom = LocalDate.now()
            }
            vilkaarsresultat {
                vilkaar = Vilkaar.FO_883_2004_ART12_1
                isOppfylt = false
                begrunnelse(Utsendt_arbeidstaker_begrunnelser.IKKE_VESENTLIG_VIRKSOMHET.kode)
            }
            vilkaarsresultat {
                vilkaar = Vilkaar.FO_883_2004_ART12_2
                isOppfylt = false
            }
        }

        val brevData = BrevDataAvslagYrkesaktiv(BrevbestillingDto(), "Z999999").apply {
            arbeidsland = Landkoder.AT.beskrivelse
            hovedvirksomhet = AvklartVirksomhet("Test AS", null, null, Yrkesaktivitetstyper.LOENNET_ARBEID)
            anmodningsperiodeSvar = AnmodningsperiodeSvar()
            yrkesaktivitet = Yrkesaktivitetstyper.LOENNET_ARBEID
            art16Vilkaar = vilkaarsresultat16_1
        }
        val spy = spyk(AvslagYrkesaktivMapper())


        val xml = spy.mapTilBrevXML(fellesType, navFelles, behandling, resultat, brevData)


        xml shouldMatch """(?s)\<\?xml version="\d\.\d+" .*>\n.*"""
    }

    @Test
    fun `mapTilBrevXML med oppfylt Art16 og anmodningsperiode bruker anmodningsperiode`() {
        val behandling = Behandling.forTest()
        val spy = spyk(AvslagYrkesaktivMapper())

        val vilkaar16_1_oppfylt = vilkaarsresultatForTest {
            vilkaar = Vilkaar.FO_883_2004_ART16_1
            isOppfylt = true
            begrunnelse(Anmodning_begrunnelser.ERSTATTER_EN_ANNEN_UNDER_5_AAR.kode)
        }

        val brevData = BrevDataAvslagYrkesaktiv(BrevbestillingDto(), "Z999999").apply {
            arbeidsland = Landkoder.ES.beskrivelse
            hovedvirksomhet = AvklartVirksomhet("Test AS", null, null, Yrkesaktivitetstyper.LOENNET_ARBEID)
            anmodningsperiodeSvar = lagAnmodningsperiodeSvarAvslag()
            yrkesaktivitet = Yrkesaktivitetstyper.LOENNET_ARBEID
            art16Vilkaar = vilkaar16_1_oppfylt
        }

        val resultat = Behandlingsresultat.forTest {
            lovvalgsperiode {
                lovvalgsland = Land_iso2.NO
                fom = LocalDate.now()
                tom = LocalDate.now()
            }
            vilkaarsresultat {
                vilkaar = Vilkaar.FO_883_2004_ART12_1
                isOppfylt = false
                begrunnelse(Utsendt_arbeidstaker_begrunnelser.IKKE_VESENTLIG_VIRKSOMHET.kode)
            }
        }


        val xml = spy.mapTilBrevXML(fellesType, navFelles, behandling, resultat, brevData)


        xml shouldMatch """(?s)\<\?xml version="\d\.\d+" .*>\n.*"""
    }

    @Test
    fun `mapTilBrevXml kan mappe alle kodeverksverdier for Art16_1 avslag`() {
        val spy = spyk(AvslagYrkesaktivMapper())
        val brevdata = BrevDataAvslagYrkesaktiv(BrevbestillingDto(), "")
        val begrunnelser = VilkaarbegrunnelseFactoryTest().lagAlleVilkaarBegrunnelser(Avslag_anmodning_begrunnelser::class)


        for (begrunnelse in begrunnelser) {
            val vilkaarsresultat = vilkaarsresultatForTest {
                begrunnelseKoder.add(begrunnelse.kode)
                begrunnelseFritekst = "Fritekst"
            }
            brevdata.art16Vilkaar = vilkaarsresultat
            spy.mapArt161Avslag(Fag(), brevdata)
        }
    }
}
